py-polars/polars/utils/udfs.py

```diff
@@ -0,0 +1,165 @@
+    """Return disassembled bytecode ops, arg-repr, and arg-specific value/flag."""
+    try:
+        instructions = list(dis.get_instructions(function))
+        idx_last = len(instructions) - 1
+        return [
+            _inst(inst.opname, inst.argrepr, inst.argval)
+            for idx, inst in enumerate(instructions)
+            if (idx, inst.opname) not in ((0, "RESUME"), (idx_last, "RETURN_VALUE"))
+        ]
+    except TypeError:
+        return []  # can't disassemble non-native functions (eg: numpy/etc)
+
+
+def _inst(opname: str, argrepr: str, argval: str) -> tuple[str, str, str]:
+    """Rewrite older opcodes using py 3.11 'BINARY_OP' instead."""
+    if sys.version_info < (3, 11) and opname in _BINARY_OPCODES:
+        argrepr = _BINARY_OPCODES[opname]
+        opname = "BINARY_OP"
+    return opname, argrepr, argval
+
+
+def _is_inefficient(ops: list[tuple[str, str, Any]], apply_target: str) -> bool:
+    """Determine if bytecode indicates only simple binary ops and/or comparisons."""
+    simple_ops = _SIMPLE_FRAME_OPS if apply_target == "frame" else _SIMPLE_EXPR_OPS
+    return bool(ops) and all(op in simple_ops for op, _argrepr, _argval in ops)
+
+
+def _rewrite_as_polars_expr(
+    ops: list[tuple[str, str, Any]], col: str | None, apply_target: str
+) -> str | None:
+    """Take postfix opcode stack and translate to native polars expression."""
+    # const / unchanged
+    if len(ops) == 1:
+        return _expr(ops[0][1], col)
+    elif len(ops) >= 3:
+        if apply_target == "expr":
+            stack = []  # type: ignore[var-annotated]
+            for op in ops:
+                stack.append(
+                    op[1]
+                    if op[0].startswith("LOAD_")
+                    else (
+                        (stack.pop(), _op(*op))
+                        if op[0] == "UNARY_NOT"
+                        else (stack.pop(-2), _op(*op), stack.pop(-1))
+                    )
+                )
+            return _expr(stack[0], col)  # type: ignore[arg-type]
+
+    # TODO: frame apply (account for BINARY_SUBSCR)
+    # TODO: series apply (rewrite col expr as series)
+    return None
+
+
+def _simple_signature(function: Callable[[Any], Any]) -> bool:
+    try:
+        return len(signature(function).parameters) == 1
+    except ValueError:
+        return False
+
+
+def warn_on_inefficient_apply(
+    function: Callable[[Any], Any], columns: list[str], apply_target: str
+) -> None:
+    """Generate ``PolarsInefficientApplyWarning`` on poor usage of ``apply`` func."""
+    # only consider simple functions with a single col/param
+    if _simple_signature(function):
+        # fast function disassembly to get bytecode ops
+        ops = _get_bytecode_ops(function)
+
+        # if ops indicate a trivial function that should be native, warn about it
+        if _is_inefficient(ops, apply_target):
+            col = (columns and columns[0]) or None
```

Comment:
what's the use case for this? when can it be `None`?

Reply:
this doesn't look quite right
```python
In [2]: (            pl.DataFrame({"col1": [[1, 2, 3]]})
   ...:             .with_columns(
   ...:                 [
   ...:                     pl.col("col1").list.eval(
   ...:                         pl.element().apply(lambda x: x, return_dtype=pl.Int64)
   ...:                     )
   ...:                 ]
   ...:             ))
> /home/marcogorelli/polars-dev/py-polars/polars/utils/udfs.py(118)_rewrite_as_polars_expr()
-> if len(ops) == 1:
(Pdb) c
<ipython-input-2-2c8912ee82d0>:5: PolarsInefficientApplyWarning: Expr.apply is significantly slower than the native expressions API.
Only use if you absolutely CANNOT implement your logic otherwise.
In this case, you can replace your function with an expression:
-  pl.col("").apply(<function <lambda> at 0x7f729daa68c0>)
+  x
```

I think we can just exclude the `None` case?

Reply:
Got an improvement for this one...